### PR TITLE
IGNITE-19591 Implemented usage of shared RocksDB instances for multiple tables.

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-19591

### CODE REVIEW

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/instance/SharedRocksDbInstance.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+    private ColumnFamily getOrCreateColumnFamily(byte[] cfName, ByteArray name) {
+        return sortedIndexCfs.computeIfAbsent(name, unused -> {
+            ColumnFamilyDescriptor cfDescriptor = new ColumnFamilyDescriptor(cfName, sortedIndexCfOptions(cfName));
+
+            ColumnFamily columnFamily;
+            try {
+                columnFamily = ColumnFamily.create(db, cfDescriptor);
+            } catch (RocksDBException e) {
+                throw new StorageException("Failed to create new RocksDB column family: " + new String(cfDescriptor.getName(), UTF_8), e);
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/instance/SharedRocksDbInstance.java:233]: Maybe create a helper common method to get the name from **ColumnFamilyDescriptor**?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbTableStorage.java [hunk 2]
```diff
@@ -239,7 +240,9 @@ private void destroyTableData() {
 
             rocksDb.db.write(DFLT_WRITE_OPTS, writeBatch);
         } catch (RocksDBException e) {
-            //TODO Handle.
+            String message = format("Failed to destroy table data. [tableId={}]", getTableId());
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbTableStorage.java:243]: U can use `org.apache.ignite.internal.storage.StorageException#StorageException(int, java.lang.String, java.lang.Throwable, java.lang.Object...)`

File: modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvPartitionStorageTest.java [hunk 3]
```diff
@@ -803,30 +802,25 @@ void readByTimestampWorksCorrectlyIfNoUncommittedValueExists() {
...
-     * Tests that changed {@link MvPartitionStorage#lastAppliedIndex()} can be successfully read and that it's returned from
-     * {@link MvPartitionStorage#persistedIndex()} after the {@link MvPartitionStorage#flush()}.
+     * Tests that changed {@link MvPartitionStorage#lastAppliedIndex()} can be successfully read back.
      */
     @Test
     void testAppliedIndex() {
         storage.runConsistently(locker -> {
             assertEquals(0, storage.lastAppliedIndex());
             assertEquals(0, storage.lastAppliedTerm());
-            assertEquals(0, storage.persistedIndex());
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/AbstractMvPartitionStorageTest.java:814]: I removed the method

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/ColumnFamilyUtils.java [hunk 4]
```diff
@@ -67,24 +93,135 @@ static ColumnFamilyType fromCfName(String cfName) {
...
-     *
-     * @param cfName Column Family name.
-     *
-     * @see #sortedIndexCfName
+     * Creates an {@link org.rocksdb.AbstractComparator} instance to compare keys in column family with name {@code cfName}.
+     * Please refer to {@link #sortedIndexCfName(List)} for the details of the CF name encoding.
      */
-    static int sortedIndexId(String cfName) {
-        return Integer.parseInt(cfName.substring(SORTED_INDEX_CF_PREFIX.length()));
+    public static RocksDbBinaryTupleComparator comparatorFromCfName(byte[] cfName) {
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/ColumnFamilyUtils.java:134]: Sure

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java [hunk 5]
```diff
@@ -170,14 +173,18 @@ public RocksDbTableStorage createMvTable(
...
         assert dataRegion != null : "tableId=" + tableId + ", dataRegion=" + tableDescriptor.getDataRegion();
 
-        Path tablePath = storagePath.resolve(TABLE_DIR_PREFIX + tableId);
-
-        try {
-            Files.createDirectories(tablePath);
-        } catch (IOException e) {
-            throw new StorageException("Failed to create table store directory for table: " + tableId, e);
-        }
+        SharedRocksDbInstance sharedInstance = sharedInstances.computeIfAbsent(tableDescriptor.getDataRegion(), name -> {
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java:176]: We should revisit this whole data region thing in rocksdb anyway, I think I'll create a JIRA for it

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/instance/SharedRocksDbInstance.java (hunk 1 above)
Comment (ibessonov) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/instance/SharedRocksDbInstance.java:233]: I'll do it

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbTableStorage.java (hunk 2 above)
Comment (ibessonov) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbTableStorage.java:243]: Ok

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/ColumnFamilyUtils.java [hunk 6]
```diff
@@ -92,6 +90,10 @@ public static ColumnFamilyType fromCfName(String cfName) {
         }
     }
 
+    public static String stringName(byte[] cfName) {
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/ColumnFamilyUtils.java:93]: Missing javadoc, I think a more appropriate name `toStingName`

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/instance/SharedRocksDbInstanceCreator.java [hunk 7]
```diff
@@ -189,7 +188,7 @@ private List<ColumnFamilyDescriptor> getExistingCfDescriptors(Path path) throws
 
     @SuppressWarnings("resource")
     private ColumnFamilyOptions createCfOptions(byte[] cfName, Path path) {
-        var utf8cfName = new String(cfName, UTF_8);
+        String utf8cfName = ColumnFamilyUtils.stringName(cfName);
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/instance/SharedRocksDbInstanceCreator.java:191]: Use static import

**Pull request merged**